cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,256 @@
+	}
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+	// in case of local etcd
+	if cfg.Etcd.External != nil {
+		fmt.Println("[control-plane-join] adding etcd in external mode")
+		return nil
+	}
+
+	// Adds a new etcd instance; in order to do this the new etcd instance should be "announced" to
+	// the existing etcd members before being created.
+	// This operation must be executed after kubelet is already started in order to minimize the time
+	// between the new etcd member is announced and the start of the static pod running the new etcd member, because during
+	// this time frame etcd gets temporary not available (only when moving from 1 to 2 members in the etcd cluster).
+	// From https://coreos.com/etcd/docs/latest/v2/runtime-configuration.html
+	// "If you add a new member to a 1-node cluster, the cluster cannot make progress before the new member starts
+	// because it needs two members as majority to agree on the consensus. You will only see this behavior between the time
+	// etcdctl member add informs the cluster about the new member and the new member successfully establishing a connection to the 	// existing one."
+	if err := etcdphase.CreateStackedEtcdStaticPodManifestFile(client, kubeadmconstants.GetStaticPodDirectory(), cfg.NodeRegistration.Name, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint); err != nil {
+		return errors.Wrap(err, "error creating local etcd static pod manifest file")
+	}
+
+	return nil
+}
+
+func runUploadConfigPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlaneJoinData)
+	if !ok {
+		return errors.New("control-plane-join phase invoked with an invalid data struct")
+	}
+
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+	client, err := data.ClientSetFromFile(kubeConfigFile)
+	if err != nil {
+		return errors.Wrap(err, "couldn't create Kubernetes client")
+	}
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	if err := uploadconfigphase.UploadConfiguration(cfg, client); err != nil {
+		return errors.Wrap(err, "error uploading configuration")
+	}
+
+	return nil
+}
+
+func runMarkControlPlanePhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlaneJoinData)
+	if !ok {
+		return errors.New("control-plane-join phase invoked with an invalid data struct")
+	}
+
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+	client, err := data.ClientSetFromFile(kubeConfigFile)
+	if err != nil {
+		return errors.Wrap(err, "couldn't create Kubernetes client")
+	}
+	cfg, err := data.InitCfg()
+	if err != nil {
+		return err
+	}
+
+	if err := markcontrolplanephase.MarkControlPlane(client, cfg.NodeRegistration.Name, cfg.NodeRegistration.Taints); err != nil {
+		return errors.Wrap(err, "error applying control-plane label and taints")
+	}
+
+	return nil
+}
+
+func runControlPlaneJoinSubphase(component string) func(c workflow.RunData) error {
```

Comment:
rename this to `runControlPlaneJoinComponentSubphase`

Reply:
removed because it is already implemented in control-plane-prepare phase